src/components/Carousel.tsx

```diff
@@ -1,18 +1,323 @@
-import React from 'react';
-import './Carousel.scss';
-
-const Carousel: React.FC = () => (
-  <div className="Carousel">
-    <ul className="Carousel__list">
-      <li><img src="./img/1.png" alt="1" /></li>
-      <li><img src="./img/1.png" alt="2" /></li>
-      <li><img src="./img/1.png" alt="3" /></li>
-      <li><img src="./img/1.png" alt="4" /></li>
-    </ul>
-
-    <button type="button">Prev</button>
-    <button type="button">Next</button>
-  </div>
-);
+import React, { useEffect, useState } from 'react';
+import '../styles/Carousel.scss';
+
+type Props = {
+  images: string[];
+};
+
+const Carousel: React.FC<Props> = ({ images }) => {
+  const [toolTipValue, setToolTipValue] = useState<number>(1);
+  const [itemWidth, setItemWidth] = useState<number>(130);
+  const [itemNumber, setItemNumber] = useState<number>(3);
+  const [step, setSteps] = useState<number>(3);
+  const [animationDuration, setAnimationDuration] = useState<number>(300);
+  const [infiniteRotation, setInfiniteRotation] = useState<boolean>(false);
+  const [position, setPosition] = useState<number>(0);
+  const [frameWidth, setFrameWidth] = useState<number>(itemNumber * itemWidth);
+
+  useEffect(() => {
+    const toolTips = document.querySelectorAll('.toolTip');
+
+    if (toolTips) {
+      const sliders = document.querySelectorAll('.slider');
+
+      sliders.forEach((slider, index) => {
+        const value = Number(slider.getAttribute('value'));
+        const toolPosition = ((value - Number(slider.getAttribute('min')))
+          / (Number(slider.getAttribute('max'))
+            - Number(slider.getAttribute('min'))))
+          * (Number(slider.clientWidth) - Number(toolTips[index].clientWidth));
+
+        (toolTips[index] as HTMLElement).style.left = `${toolPosition}px`;
+      });
+    }
+  }, [toolTipValue]);
+
+  useEffect(() => {
+    setFrameWidth(itemNumber * itemWidth);
+    setPosition(0);
+  }, [itemNumber, itemWidth]);
+
+  useEffect(() => {
+    if (infiniteRotation) {
+      setPosition(0);
+    }
+  }, [infiniteRotation]);
+
```

Comment:
i think you're missing another useEffect

the usecase it covers is: when you have 5 items in frame && step is 4
and you change items in frame to something that's below step, step is not re-applied -> images pass-by.

Try it on your current demo.

```suggestion

  useEffect(() => {
    if (itemNumber < step) {
      setSteps(itemNumber)
    }
  }, [itemNumber, step])
```

Reply:
I've looked into the issue and did not see any difference.
Need a second pair of goggles 🥽 😄